client/lib/streamframer/framer.go

```diff
@@ -0,0 +1,310 @@
+}
+
+// run is the internal run method. It exits if Destroy is called or an error
+// occurs, in which case the exit channel is closed.
+func (s *StreamFramer) run() {
+	var err error
+	defer func() {
+		s.l.Lock()
+		s.running = false
+		s.err = err
+		s.l.Unlock()
+		close(s.exitCh)
+	}()
+
+OUTER:
+	for {
+		select {
+		case <-s.shutdownCh:
+			break OUTER
+		case <-s.flusher.C:
+			// Skip if there is nothing to flush
+			s.l.Lock()
+			if s.f.IsCleared() {
+				s.l.Unlock()
+				continue
+			}
+
+			// Read the data for the frame, and send it
+			s.f.Data = s.readData()
+			err = s.send(&s.f)
+			s.f.Clear()
+			s.l.Unlock()
+			if err != nil {
+				return
+			}
+		case <-s.heartbeat.C:
+			// Send a heartbeat frame
+			if err = s.send(HeartbeatStreamFrame); err != nil {
+				return
+			}
+		}
+	}
+
+	s.l.Lock()
+	if !s.f.IsCleared() {
+		s.f.Data = s.readData()
+		err = s.send(&s.f)
+		s.f.Clear()
+	}
+	s.l.Unlock()
+}
+
+// send takes a StreamFrame, encodes and sends it
+func (s *StreamFramer) send(f *StreamFrame) error {
+	sending := *f
+	f.Data = nil
+
+	select {
+	case s.out <- &sending:
+		return nil
+	case <-s.exitCh:
+		return nil
+	}
+}
+
+// readData is a helper which reads the buffered data returning up to the frame
+// size of data. Must be called with the lock held. The returned value is
+// invalid on the next read or write into the StreamFramer buffer
+func (s *StreamFramer) readData() []byte {
+	// Compute the amount to read from the buffer
+	size := s.data.Len()
+	if size > s.frameSize {
+		size = s.frameSize
+	}
+	if size == 0 {
+		return nil
+	}
+	d := s.data.Next(size)
+	return d
+}
+
+// Send creates and sends a StreamFrame based on the passed parameters. An error
+// is returned if the run routine hasn't run or encountered an error. Send is
+// asynchronous and does not block for the data to be transferred.
+func (s *StreamFramer) Send(file, fileEvent string, data []byte, offset int64) error {
+	s.l.Lock()
+	defer s.l.Unlock()
+
+	// If we are not running, return the error that caused us to not run or
+	// indicated that it was never started.
+	if !s.running {
+		if s.err != nil {
+			return s.err
+		}
+
+		return fmt.Errorf("StreamFramer not running")
+	}
+
+	// Check if not mergeable
+	if !s.f.IsCleared() && (s.f.File != file || s.f.FileEvent != fileEvent) {
+		// Flush the old frame
+		s.f.Data = s.readData()
+		select {
+		case <-s.exitCh:
+			return nil
+		default:
+		}
+		err := s.send(&s.f)
+		s.f.Clear()
+		if err != nil {
+			return err
+		}
+	}
+
+	// Store the new data as the current frame.
+	if s.f.IsCleared() {
+		s.f.Offset = offset
+		s.f.File = file
+		s.f.FileEvent = fileEvent
+	}
+
+	// Write the data to the buffer
+	s.data.Write(data)
+
+	// Handle the delete case in which there is no data
+	force := false
+	if s.data.Len() == 0 && s.f.FileEvent != "" {
+		force = true
+	}
+
+	// Flush till we are under the max frame size
+	for s.data.Len() >= s.frameSize || force {
+		// Clear
+		if force {
```

Comment:
This logic is a bit confusing, as force will be set to false the first time through this loop. Is this the intended behaviour?